src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,237 @@
+        currentNode = head;
+        if (head == null) {
+            index = 0;
+            head = newNode;
+            last = newNode;
+        } else {
+            while (currentNode.next != null) {
+                currentNode = currentNode.next;
+            }
+            currentNode.next = newNode;
+            newNode.prev = currentNode;
+            last = newNode;
+            index++;
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        Node<T> newNode = new Node<>(value);
+        if (!isEmpty() && index == 0) {
+            head.prev = newNode;
+            newNode.next = head;
+            head = newNode;
+            this.index++;
+        } else if (!isEmpty() && index == size()) {
+            add(value);
+        } else if (isEmpty() && index == 0) {
+            add(value);
+        } else {
+            currentNode = getNode(index);
+            prevNode = currentNode.prev;
+            prevNode.next = newNode;
+            newNode.next = currentNode;
+            currentNode.prev = newNode;
+            newNode.prev = prevNode;
+            this.index++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        currentNode = head;
+        if (isValidIndex(index) && index == 0) {
+            return currentNode.item;
+        } else if (index == this.index) {
+            currentNode = last;
+            return currentNode.item;
+        }
+        currentNode = getNode(index);
+        return currentNode.item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        T item;
+        if (isValidIndex(index) && index == 0) {
+            item = head.item;
+            head.item = value;
+        } else if (index == this.index) {
+            item = last.item;
+            last.item = value;
+        } else {
+            currentNode = getNode(index);
+            item = currentNode.item;
+            currentNode.item = value;
+        }
+        return item;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        T value;
+        if (size() == 1 && index == 0) {
+            value = head.item;
+            removeHeadAndLastNode();
+        } else if (isValidIndex(index) && index == 0) {
+            value = head.item;
+            head = head.next;
+            head.prev = null;
+            currentNode = head;
+            this.index--;
+        } else if (index == this.index) {
+            value = last.item;
+            last = last.prev;
+            last.next = null;
+            currentNode = last;
+            this.index--;
+        } else {
+            currentNode = getNode(index);
+            value = currentNode.item;
+            removeNode();
+        }
+        return value;
```

Comment:
```suggestion
        Node<T> node = getNode(index);
        return unlink(node);
```
all this logic should be in unlink